onchain/rollups/test/foundry/portals/ERC1155BatchPortal.t.sol

```diff
@@ -80,6 +82,24 @@ contract ERC1155BatchPortalTest is Test {
         bob = vm.addr(3);
     }
 
+    function testSupportsInterface(bytes4 _randomInterfaceId) public {
```

Comment:
libraries cannot inherit. So that means we won't be able to use `assertTrue`, `assertFalse` and `vm.assume` in the library, because those need to inherit from `forge-std/Test.sol`. So a library won't save much lines of codes.
We could consider making it as a helper contract rather than a library. So the helper contract can inherit from `forge-std/Test.sol` and then in turn the test contract inherits from the helper contract. Then there's also a 3rd parameter for the helper function: the fuzzy parameter `_randomInterfaceId` from the test function. That doesn't seem too clean either.
Maybe the current way is easier to read, but modifying with `assertTrue` and `assertFalse`. What do you think?

Reply:
Ok, I agree! So let's create an abstract contract under `test/foundry/util` called `ERC165Test` which inherits from `forge-std/Test.sol:Test` and defines:

- a public function called `testSupportsInterface`;
- a public function called `testDoesNotSupportRandomInterface`, which takes an interface ID as a fuzzy parameter;
- a public function called `testDoesNotSupportInvalidInterface`;
- a virtual function called `getERC165Contract`, which returns an ERC-165-compliant contract;
- a virtual function called `getSupportedInterfaceIDs`, which returns an array of interface IDs;

So, test contracts that deal with contracts that implement ERC-165 can inherit from `ERC165Test` and define the `getERC165Contract` and `getSupportedInterfaceIDs` functions. The `test*` functions will be called by `forge test` automatically.

The `testSupportsInterface` function would call `getERC165Contract()` and `getSupportedInterfaceIDs()` and iterate over the interface IDs, checking that the provided `IERC165` contract implements all of them. Additionally, it would also check if it implements `IERC165`.

And the `testDoesNotSupportInterface` function would also call both functions, but `vm.assume` that the provided interface ID is NOT in the array of supported interface IDs. It would then check if the `IERC165` contract does NOT support the provided interface ID.

Finally, the `testDoesNotSupportInvalidInterface` function would only check that the `IERC165` contract does not support `0xffffffff`.

Reply:
I think this is over-complicated, don't you think? The way it currently is is easy to read already